shopinvader/components/service_context_provider.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2021 ACSONE SA/NV
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+from odoo.addons.component.core import Component
+from odoo.addons.shopinvader.utils import get_partner_work_context
+
+
+class ShopinvaderServiceContextProvider(Component):
+    _name = "shopinvader.service.context.provider"
+    _inherit = "base.rest.service.context.provider"
+    _collection = "shopinvader.backend"
+
+    def _find_partner(self, backend, partner_email):
+        partner_domain = [
+            ("partner_email", "=", partner_email),
+            ("backend_id", "=", backend.id),
+        ]
+        return self.env["shopinvader.partner"].search(partner_domain, limit=2)
+
+    def _validate_partner(self, backend, partner):
+        return backend._validate_partner(partner)
+
+    def _get_shopinvader_session(self):
+        # HTTP_SESS are data that are store in the shopinvader session
+        # and forwarded to odoo at each request
+        # it allow to access to some specific field of the user session
+        # By security always force typing
+        # Note: rails cookies store session are serveless ;)
+        return {
+            "cart_id": int(self.request.httprequest.environ.get("HTTP_SESS_CART_ID", 0))
+        }
+
+    def _get_shopinvader_partner(self):
+        """Get the partner requesting the api
+
+        At this stage, the partner returned by this method must be the shopinvader
+        partner...
+        """
+        return self.env["shopinvader.partner"].browse()
+
+    def _get_backend(self):
+        """Get the requested shopinvader backend instance"""
+        tech_name = self.request.httprequest.environ.get("HTTP_WEBSITE_ID")
```

Comment:
This means that whoever has the website ID can bombard Odoo w/ fake orders?
The tech name will be very likely easier to determine than an api key.
This implies that the recommended (if not required) way is to always proxy the calls and firewall `/shopinvader` path.
What about the future when we won't have locomotive proxying calls?
Do you already have a clear path in miind? If yes, it would be nice to see it the docs/readme.

Just asking :wink:

Reply:
WEBSITE_ID is the tech name....
This header should be added by the reverse proxy or the lib that manage the call to the api from the website. We've to find a way to specify the target backend for the calls to the API for unauthenticated users... motivations can be found [here](https://github.com/shopinvader/odoo-shopinvader/pull/1031#discussion_r673100157)

Reply:
@simahawk this does not replace authentication, via api key or jwt, it is only a mechanism to address the shopinvader backend. So this does not imply proxying is necessary, I think ?

If we worry that a user would guess the name of another shopinvader backend, we can recommend to use a long random string.

In any case even if a user guesses it, it can only do things that he/she is authorized to do.
And for public routes (such as the anonymous cart) the spamming problem already exists with locomotive, and the only solutions I see for that are captchas and/or rate limiting.

Do you see other risks ?

Reply:
Of course I'm talking about anon users. For auth users this is not a problem.

Actually with current implementation (with locomotive) is a bit better: Locomotive is hosted on it's own server and connections can be limited to that. With this implementation you can potentially call this from the browser itself.
I know the issue is kind of already there but with this change I feel it's even more required to highlight it.

So yeah, I think we should write clearly and in a prominent way that:

1. the webserver / reverse proxy (eg: nginx) should strip out any key matching this one and inject the good one (possibly provide a basic configuration as example)
2. the key should be a random one and not just a tech name -> I would keep `tech_name` for the env configuration (to assign the key) and call this field `website_hash` or `website_unique_key` or something similar

Reply:
@simahawk implemented in e1565cf